lib/apiservers/engine/backends/archive.go

```diff
@@ -0,0 +1,553 @@
+			return fmt.Errorf("item not ArchiveWriter")
+		}
+
+		aw, _ = item.(*ArchiveWriter)
+		return nil
+	}
+
+	// Find the prefix for the final destination.  Final destination is the combination of container destination path
+	// and the asset's name.  For example,
+	//
+	//	container destination path =	/
+	//	asset name =					mnt/A/file.txt
+	//	mount 1	=						/mnt/A
+	//	mount prefix =					/mnt/A
+	//
+	// In the above example, mount prefxi can only be determined by combining both the container destination path and
+	// the asset name, as the final destination includes a mounted volume.
+	combinedPath := path.Join(containerDestPath, assetName)
+	prefix := patricia.Prefix(combinedPath)
+	err = wm.prefixTrie.VisitPrefixes(prefix, findPrefix)
+	if err != nil {
+		log.Errorf(err.Error())
+		return nil, fmt.Errorf("Failed to find a node for prefix %s: %s", containerDestPath, err.Error())
+	}
+
+	if aw == nil {
+		return nil, fmt.Errorf("No archive writer found for container destination %s and asset name %s", containerDestPath, assetName)
+	}
+
+	return aw, nil
+}
+
+// WriterForAsset takes a destination path and subpath of the archive data and returns the
+// appropriate writer for the two.  It's intention is to solve the case where there exist
+// a mount point and another mount point within the first mount point.  For instance, the
+// prefix map can have,
+//
+//		R/W -					/
+//		mount 1 -				/mnt/a
+//		mount 2 -				/mnt/a/b
+//
+//		case 1:
+//				containerDestPath -			/mnt/a
+//				archive header source -		b/file.txt
+//
+//			The correct writer would be the one corresponding to mount 2.
+//
+//		case 2:
+//				containerDestPath -			/mnt/a
+//				archive header source -		file.txt
+//
+//			The correct writer would be the one corresponding to mount 1.
+//
+//		case 3:
+//				containerDestPath -			/
+//				archive header source -		mnt/a/file.txt
+//
+//			The correct writer would be the one corresponding to mount 1
+//
+// As demonstrated above, the mount prefix and writer cannot be determined with just the
+// container destination path.  It must be combined with the actual asset's name.
+func (wm *ArchiveStreamWriterMap) WriterForAsset(proxy VicContainerProxy, cid, containerDestPath string, assetHeader tar.Header) (io.WriteCloser, error) {
+	defer trace.End(trace.Begin(assetHeader.Name))
+
+	var err error
+	var streamWriter io.WriteCloser
+
+	aw, err := wm.FindArchiveWriter(containerDestPath, assetHeader.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// Perform the lazy initialization here.
+	if aw.writer == nil {
+		// lazy initialize.
+		log.Debugf("Lazily initializing import stream for %s", aw.mountPoint.Destination)
+		var deviceID string
+		if aw.mountPoint.Destination == "/" {
+			// Special case. / refers to container VMDK and not a volume vmdk.
+			deviceID = cid
+		} else {
+			deviceID = aw.mountPoint.Name
+		}
+		streamWriter, err = proxy.ArchiveImportWriter(context.Background(), "container", deviceID, aw.filterSpec)
```

Comment:
This method is using "container" as the store. Are they using same store for volume mount and container disk?
btw, can we have constants for all supported store type?